src/coreclr/src/jit/compiler.h

```diff
@@ -8285,7 +8276,10 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
         return false;
     }
 
-    bool compSupports(InstructionSet isa) const
+    // Answer the question: Is a particular ISA supported?
+    // Use this api when asking the question so that future
+    // ISA questions can be asked correctly.
+    bool compSupportsNoReporting(InstructionSet isa) const
```

Comment:
So this one asks "is the ISA supported" without changing any flags, etc
`compSupports` asks but also marks the ISA as being used
`compSupportsOptional` asks and marks it as used if it is considered part of the baseline?

Reply:
Not quite. 

compSupportsNoReporting asks is the ISA supported without reporting anything to the VM side. Its extremely dangerous to use, and really only valuable for setting up the set of instruction sets useable internally in the jit. Auditing usage of this is quite easy. If the result of this has any direct impact on codegen, it is a bug.

compSupportsOptional asks is the ISA supported. If it returns true, it will report the instruction set usage to the VM, but if false, it does no reporting. This function is only to be used when returning false will not cause the JIT to generate code that will not work correctly if the code is used on a machine which does support the instruction set. Use of this function should be audited with significant care, as incorrect usage of the return value may cause use of illegal instructions or other incorrect behavior at runtime. Unfortunately, I couldn't see a way to usefully do CPU feature usage detection without implementing an api like this, or more significant restructuring of the jit codebase.

compSupports asks is the ISA supported. The fact that the jit asked such a question is recorded and the generated code will not be used on hardware which does not exactly match the result of the call. For instance, if compSupports(Avx) is called, then if it returns false, the generated code will never be executed on hardware which supports Avx. This is a safe api to use at all times. Over-usage of this api may cause unnecessary prevention of functions from being used at runtime.

Reply:
I wonder if we could come up with names that make their usage more clear as to the intended usage 🤔

Reply:
I'm notoriously bad at naming (the names make sense to me...), so I'd welcome suggestions, but really I'd rather someone from the JIT team chime in on that, as they are the ones that will have to maintain them going forward, and I want it to make sense to them. @dotnet/jit-contrib

Reply:
I totally agree with @tannergooding - these are extremely confusing, and the idea of having anything that's "extremely dangerous to use" in the JIT just sounds like a bad idea. I think we really need to think this through carefully and figure out how to add the appropriate validation.

This statement is quite confusing with the triple negation: "This function is only to be used when returning false will not cause the JIT to generate code that will not work correctly if the code is used on a machine which does support the instruction set." I had to look through the code before I understood what it meant. AFAICT the only reason for this is to prevent mixing code with different ideas of the size of `Vector<T>`.

How about:
- `bool compCanUse(InstructionSet isa)` - this tells the JIT that it's allowed to use a given ISA, but reports no usage to the VM.
- `bool compSetUsed(InstructionSet isa)` - this checks whether the JIT is allowed to use a given ISA and if so, tells the VM that it is doing so.
- `void compDisallow(InstructionSet isa)` - this tells the VM not to use this code on a machine that supports the given isa. This would handle the `Vector<T>` case.

Reply:
I don't feel that those semantics are viable. The requirement of those semantics would be to take every use of compSupports that I left in the codebase (which is quite a few) and replace it with, 

```
bool canUse = compSetUsed(isa);
if (!canUse)
    compDisallow(isa);
```

The issue here is not the mixing of code with different ideas of the size of `Vector<T>`, but to properly handle cases like...

```
Vector256<int> DoSomethingAmazing(Vector256<int> input)
{
    if (Avx.IsSupported)
    {
        return Avx.SomeSeriesOfIntrinsics(input);
    }
    else
    {
        throw new Exception("Fail spectacularly");
    }
}
```

If we compile the above with only SSE42 support specified, then the method body in this design will simplify to `throw new Exception("Fail spectacularly");`, but when we run the application on a machine supporting Avx it should do a set of intrinsic things.

Reply:
In general, my belief is that somethjing like `compCanUse` is only to be used in very restricted locations, and use throughput the jit would be dangerous, but could easily be restricted to the 2-3 places in the jit where it is valuable, and thus as it is essentially forbidden to be used. It matches exactly with my `compSupportsNoReporting`. I don't really dislike the name, except that it looks like its an api that should be used by much of the codebase, and it really isn't. 

`compSetUsed` has exactly the semantic of the compSupportsOptional api that I propose, and I believe it is the most difficult to review and use correctly. The problem with it is that a return value of false may do something unfortunate. In my comment above I discuss how it could change program behavior in a way that wouldn't cause a fault in the EE, and would just throw a perfectly normal managed exception. But really, there is no predicting what not reporting a false in many places might do.

My goal was to keep the vast majority of the codebase to using `compSupports`, which would continue to have the same semantic as it does today. Sure, it does extra reporting, but it will never generate incorrect code or behave wrong if `compSupports` is used.

Reply:
I believe I understand the desired semantics. There are two APIs, each taking an `isa' argument.
- The first API:
  - returns `true` if `isa` is supported, and records that it must be supported.
  - returns `false` if `isa` is not supported, and records that it must *not* be supported. The JIT is free to generate code that depends on the associated `isSupported` call returning `false` (whether not such a call actually exists in the code).
- The second API:
  - returns `true` if `isa` is supported and records that it must be supported
  - if it returns `false`, the JIT must generate semantically equivalent code as for the `true` case, such that the code will have the same behavior in either case.

I am of the opinion that these need to have names that communicate these semantics more clearly. I was thinking `compEnableOrProhibit` and `compEnableOrAllow`, but I don't like these much better than `compSupports` and `compSupportsOptional`.

Perhaps others have better suggestions.

Reply:
Also, in response to:

> My goal was to keep the vast majority of the codebase to using `compSupports`, which would continue to have the same semantic as it does today.

That's not exactly true. Today, the JIT doesn't have separate notions of an alternate path that either depends on the ISA *not* being supported on the target **or** allows it to be supported on the target but perhaps with slightly less efficient code being used.

Reply:
How about this...

`compMayUseIsa(isa)`
- Informs the jit that it is allowed to use a particular `isa` in its codegen
- This usage must not change the behavior of the generated application, although it may change its performance characteristics
- It returns true if `isa` is supported, and when operating as an AOT compiler records into the R2R image that the `isa` must be supported at runtime of the generated code.
- It returns false if `isa` is not supported. This false is not recorded with the AOT compiler as the lack of use of the instruction set is not to change the behavior of the application.

Example usages:
When choosing to generate the code for a call to the Math.Round function, the JIT may choose to generate a particular optimized sequence of assembly instructions if SSE4.1 is available, but will otherwise call the Math.Round function. In this case, the compMayUseIsa function is appropriate to use.
Example usages which are not permitted:
When generating the code for Sse42.IsSupported, the compMayUseIsa function must not be used. If the application has code that uses Sse42.IsSupported, the return value of the IsSupported intrinsic is semantically meaningful, and changes in its value are visible in the state of the program. Thus compMayUseIsa must not be used.

`compUsesOrDoesNotUseIsa(isa)`
- Informs the jit that it is allowed to use a particular
- This usage may change the behavior of the generated application, as in changing the meaning of Vector<T>.Size, or the return value of Sse42.IsSupported.
- It returns true if `isa` is supported, and when operating as an AOT compiler records into the R2R image that the `isa` must be supported at runtime of the generated code.
- It returns false if `isa` is not supported, and when operating as an AOT compiler records into the R2R image that the `isa` must not be supported at runtime of the generated code.

Example usages:
When choosing the behavior of an IsSupported intrinsic, or when generating the code for a platform intrinsic. The behavior of these intrinsics varies based on the presence/absence of support, and thus the runtime must verify that the `isa` is supported exactly as when compiled.

Example usages which are discouraged (although not strictly incorrect):
When choosing to generate the code for a call to the Math.Round function, the JIT may choose to generate a particular optimized sequence of assembly instructions if SSE4.1 is available, but will otherwise call the Math.Round function. compUsesIsa should not be used for this case, as the usage of the instruction set has no bearing on the behavior of the generated code.

Reply:
Could we just have `enum compSupports(Isa)` and have it return `Unsupported`, `Supported`, or `SupportedWithCheck`? I believe that covers all the considerations and is incredibly simple to use/understand.

Basically, when jitting, the compiler is already doing the right thing, `compSupports` today return effectively just `Unsupported` (`false`) or `Supported` (`true`).
With the introduction of the AOT scenarios (R2R, Crossgen, etc), we need to consider a third state which can be represented as above:
* If the instruction set is unsupported, even behind a check, then we return `Unsupported` and we should never go down that path
* If the instruction is part of the baseline, we get `Supported` and it can be emitted without any considerations, for example `IsSupported` can be constant folded to be `true`
* If the instruction is part of the baseline encoding but not part of the baseline, we get `SupportedWithCheck`, which indicates you can only use it if behind some kind of `IsSupported` check
  * For HWIntrinsics, this represents no change in behavior. We just emit the instruction and it will fail at runtime with `#UD` (Undefined Opcode) if the user didn't do the appropriate check
  * For other code paths, like `Math.Round` this indicates we can't emit this without some corresponding `IsSupported` check being in place (we don't currently support such a check, so we treat it as `Unsupported`

Reply:
I'm not crazy about the `compMayUseIsa` - again, it fails to communicate that it conveys a contract, i.e. that when it returns false the code must execute correctly even on a target where it would have returned true.

@davidwrighton - in your discussion above you say:
> `compUsesIsa` should not be used for this case...

I presume you mean `compUsesOrDoesNotUseIsa`.

The issue I have with `compUsesOrDoesNotUseIsa` is that it seems to be asking if we're using the isa or not - as opposed to stating a contract as part of the query. What we really want it to say is `tellMeIfICanUseThisIsaAndIfNotDisableIt` and `tellMeIfICanUseThisIsaButIfNotAllowIt` but of course that's just a bit verbose ;-)

> Could we just have enum compSupports(Isa) and have it return Unsupported, Supported, or SupportedWithCheck?

I don't think that addresses the requirements that @davidwrighton has expressed - i.e. that the JIT needs to communicate which kind of support it wants, as different codegen scenarios require different answers.

Reply:
Let me restate in terms of my two favorite concepts -- legality and profitability -- and see if that clarifies or further confuses things.

Seems like there are two goals overall:
* `compUsesOrDoesNotUseIsa(isa)` [legality] to determine if some precompiled method can safely be used at runtime, interoperating correctly and providing the exact same results (though perhaps not the same performance) as if the method was jitted.
  * all `IsSupported` calls must have the same value when prejitting and jitting; so when expanding these as intrinsics during prejitting, the jit must consult the vm each time it evaluates an `IsSupported` method. The prejitter must accumulate both positive and negative results from these and and make this information available at runtime.
  * if method has vector<T> codegen, vector lengths must agree exactly. Runtime/prejitter knows this value, but doesn't know if the generated code is dependent upon it, so this needs to be communicated by the jit. This is could be encoded as an ISA dependence, as the available ISA determines the vector length, but it might be simpler to convey it separately.
* `compMayUseIsa(isa)` [profitability] to determine if more performant code could be produced internally by the jit if some ISA is available at runtime. The jit should query the vm when it reaches these decision points. Positive results should result in legality dependencies; negative results should be recorded as profitability opportunities. There might be need to be some more general formulation of this if the jit's decision making is requires multiple checks before a decision can be reached.

The legality set forms a "must" or exact dependence check; if the runtime ISA does not agree completely with the `IsSupported` ISA queries made when prejitting, or the vector length does not match, or the interally dependent codegen is not supported, then the prejitted method cannot safely be used; the method must be jitted.

If this "must check" passes the runtime can safely use the prejitted code. However, it may still choose to jit. This decision can be informed by the profitability set; these form a "may" dependence set. If the profitability dependencies are are satisfied or exceeded, there is some evidence that jitting the method may improve steady-state performance by making use of ISA features available at runtime.

Reply:
> I don't think that addresses the requirements that @davidwrighton has expressed - i.e. that the JIT needs to communicate which kind of support it wants, as different codegen scenarios require different answers.

Ah, right. I guess the scenario that doesn't cover is, if `SupportedWithChecks` was returned, then the method needs to tell the JIT that "yes, I am emitting an instruction which requires this".
Maybe that could be supported by a separate method which is simply: `compSetUsed(Isa)`. Then we clearly have: "tell me the support level the compiler has" and "let me tell the compiler I am using it"

Reply:
@AndyAyersMS I really like your rephrase of the concepts here in terms of legality and profitability, but I'm not sure if you like the names I came up with.

I'm really looking from a positive statement from a reviewer that they actually like an api name, and feel they could successful review uses of it for correctness, so I'm throwing names at the wall and hoping that one of them sticks.

How about...

`compUseIsa(isa)` for the legality check
`compUseIsaForOptimization(isa)` for the profitability check

Reply:
I completely understand "legality", but I don't understand "profitability" in this context. We aren't checking the profitability of using a given method nor are we telling anything that something is profitable to do.

We already know that using this other instruction path is profitable and we want to do it if it is legal. So, we are really just telling the compiler that we did use it for a given method so that R2R (or other scenarios) can target `X` by default and still fallback (to the JIT) if you are running on something that doesn't support `X` (at least as I understood it).

Reply:
Maybe

`compExactlyDependsOn(isa)` -- returns true/false.

If true, jit can use ISA, method body requires that ISA is present at runtime
If false, jit cannot use ISA, method body requires that ISA is not present at runtime

`compOpportunisticallyDependsOn(isa)` -- returns true/false

If true, jit can (and should) use ISA, method body requires that ISA to be present at runtime
If false, jit cannot use ISA, method body marked as one that might opportunistically benefit from being rejitted if ISA is present at runtime.

and maybe

`compExactlyDependsOnVectorTLength(len?)`

Note that method body exactly depends on Vector<T> length.

Presumably all of these are basically no-ops when jitting. Not sure if we want to express this in the jit code or have the runtime side sort it out... might be nice to track the ISA dependence even for jitted code and perhaps propagate to ETL, so we can see what methods at runtime are taking advantage of ISA features by looking at the method load events.

Here by "ISA not present" we are including cases where the runtime claims support levels lower than the actual level of the hardware, so we can test downlevel API support to some extent without having actual downlevel HW.

And probably something similar with vector length, eg runtime can artificially limit vector length despite ISA support for wider lengths.

Reply:
I like your names, and feel that they would be straightforward to replace my names. The R2R file format is already capable of performing a check to ensure that `Vector<T>` size is a specific value, and I'd just need to add another jit interface api to pass that through the system so that the proper tracking data can be emitted.

With regards to tracking the ISA dependence at runtime, that may be nice, so I'd like to have the JIT unconditionally send the data through the api, but I don't think we have enough reason to actually build the runtime logging at this time (other than into the jitdump).

@tannergooding The profitability question isn't asking, is this transformation profitable? It is asking, for a given transformation which is only applied because it is a profitable optimization, is it legal to make said optimization. Defining this name as distinct from the strict legality targeting name is the really problem here, as the concept is complex.

Reply:
I'm not so sure it is complex. We basically have three states:
* Never Supported
* Always Supported
* Supported with a check

Given those three, any codepath in the JIT can determine whether or not they can emit an instruction. You then just have to notify the compiler when you decide to take advantage of a codepath, correct?

If that is the case, it can really just be:
* `compGetSupport(Isa)` which returns one of the three states
* `compSetUsed(Isa)` which tells the compiler you've used something out of a given ISA

Reply:
@davidwrighton and I talked for quite a bit on Teams. He helped me to understand why the `I checked for support but didn't use it case` was important and why it is relevant to user code, but not to S.P.Corelib.
Namely, S.P.Corelib is heavily reviewed and we have certain standards that guarantee we have software fallbacks and the various paths are equivalent. We can't actually assume that is true for user code and so knowing that `Avx` was checked for but not used is actually important.

We then discussed my proposal for a bit and I explained some of the reasoning behind it and how I think it would work given the above. Namely, we are already doing `compSupports` checks today and these just return `true` (supported) or `false` (unsupported).
Under this model, the compiler can already track the fact that an ISA check happened and therefore that the method "could" be using something from that ISA. What it doesn't track is whether or not the compiler "actually" used something from the ISA. It also doesn't cover the scenarios like `Math.Round` where the compiler supports emitting the instruction but the target machine might not support it and this is relevant information to the JIT.

In order to cover the "was an ISA actually used", the proposal is for `compSetUsed` to be called in the emitter. Basically, if codegen actually produces an `SSE4.1` instruction, then it will mark the method as "using SSE4.1", regardless of how it got there or if a higher level check happened.
There would additionally be a few methods such as `Vector<T>.Count` which logically do a `compSupports(Avx2)` check and which then mark the method as using an ISA even though they don't emit an instruction from that ISA (there are a handful of these methods and they are largely centered around `Vector<T>`).
We already largely track the ISAs for given instructions as part of the HWIntrinsic work and so this should be minimal work to follow up on and would allow additional validation in the emitter on whether emitting an instruction is "legal" (we've had cases in the past where we emitted AVX2 instructions even though `compSupports(AVX2) == false`).

In order to cover the `Math.Round` scenario, my proposal is to extend `compSupports` to return a third state which is basically "Conditionally Supported". This gives enough information for the JIT to decide if it can use an ISA specific optimization. If the ISA is "supported" (part of the baseline) we can go down that path and the emitter will mark it as used. If it is not supported, we will have tracked the query and then `compSetUsed` be called so we can track this as a "could profit from rejit". If it is conditionally supported, we can decide to either treat it as "supported" (which will mark the method as using that ISA) or "unsupported" (method will track the query, but not mark it as used); which we do is probably based on discussion around which is a better default for the JIT.

This means that methods basically have two pieces of data associated with them:
* What ISAs did the method check for support on, which is set by `compSupports`
* What ISAs is the method actually dependent on, which is set by `compSetUsed`

Using these two pieces of data, you then have the context on a method and its potential codegen states:
* If a method queries an ISA and uses an instruction from that ISA, then you know the method is valid if the underlying hardware supports that ISA and that it needs to be rejitted otherwise
* If a method queries an ISA and doesn't use an instruction, then you know it may have semantically different behavior and must be rejitted (because we can't assume users wrote correct code like we can in S.P.Corelib)
* If a method does not query an ISA but does use an instruction from that ISA, then we can functionally treat it as the first case (method is valid if underlying hardware supports that ISA, otherwise needs rejit), but we also know that the user didn't write an `IsSupported` check (or we have a potential JIT bug where we are emitting an instruction in an invalid context)

This should be all the context required to know if a given R2R method is valid for the underlying hardware and if it should be rejitted.

Reply:
(Just noting, there wasn't any decision made; David just asked me to write up a summary of the discussion 😄)

Reply:
An interesting scenario I thought of that this wouldn't cover is if you did a `Sse41` check for `Math.Round`, the JIT said "supported with check", and we didn't emit it (we treat it as "unsupported"); but some other part of the method had its own `Sse41.IsSupported` check and did emit an `Sse41` instruction; then we would miss out on that information (which is that "some path" could have benefited from this).

This wouldn't be an issue if we treat it as "supported", but it is something to consider. Likewise, I think we could potentially miss cases with the other proposal if we only rely on `compExactlyDependsOn` and `compOpportunisticallyDependsOn`. Such as if some code path emits an `Sse41` instruction without the appropriate check.

Reply:
@tannergooding - I don't understand how the above captures the scenario where the JIT asked if XXX is supported, the answer was "no", but the code is still valid even if the answer had been "yes" (versus the case where it would be incorrect code for the "yes"/supported case).

Reply:
@CarolEidt, Could you give a small example of such a scenario in psuedo code?

Reply:
@tannergooding, @CarolEidt, also from the discussion, I am entirely unconvinced that the compSupports SupportedWithCheck return value is actually useful. It might be, but only as a possible optimization we could use in the future if we find that a true/false result is insufficient.

A scenario where there is a check for an instruction set, and the result was false, but the code should still be used is the Math.Round case, or Vector4.Dot case. More generally, while these cases are somewhat rare, they do exist. In our codebase, they can be seen in the spots where I used compSupportsOptional in my change, or where the code uses a construct like getSIMDSupportLevel() and has a viable fallback path that isn't that inefficient. Remember that R2R code isn't really required to be best in class performance, it is instead intended to be pretty good, and sometimes non-ideal output is ok if it works.

A pseudo code example might look like..
```
float ComplexMathThing(float a, float b, float c)
{
    return MathF.Round(/*Some very complex expression involving a, b, and c*/ );
}
```

For all of those sorts of uses, I suspect we'd need to have a compSupportsLike function that doesn't record that an instruction is expected to be used. I suspect it be used in exactly the same locations and way as the compSupportsOptional function that started this whole discussion. Thus I believe the actual impact of following Tanner's model of how to think about the problem is just that we might get a better assert message, and it might be easier to explain the model, but the practical effect is much the same.

I still prefer the approach that I outlined about annotating the questions that are asked, but Tanner's approach has the benefit of allowing us to add an assert to the product that verifies that an instruction set was queried for before it was used by the jit.

Reply:
> I am entirely unconvinced that the compSupports SupportedWithCheck return value is actually useful

I might be inclined to agree. The initial reasoning around this was from the perspective of `S.P.Corelib` in that we could trust all code paths were "equivalent", and therefore `IsSupported` could be a cached CPUID check.
But, if we are saying we can't trust user code and therefore we must reJIT if the method could target `AVX` and the actual hardware supports `AVX`, then it isn't important.

It might be impactful to some AOT scenarios, but not without some external switch allowing users to opt into that behavior.